tests/scanner/audit/retention_rules_engine_test.py

```diff
@@ -0,0 +1,626 @@
+
+        rules_local_path = get_datafile_path(
+            __file__,
+            'bucket_retention_test_rules_0.yaml')
+        self.scanner = retention_scanner.RetentionScanner(
+            {}, {}, mock.MagicMock(), '', '', rules_local_path)
+
+        mock_data_access = mock.MagicMock()
+        mock_data_access.scanner_iter.side_effect = _mock_gcp_resource_iter_0
+        mock_service_config = mock.MagicMock()
+        mock_service_config.model_manager = mock.MagicMock()
+        mock_service_config.model_manager.get.return_value = (
+            mock.MagicMock(), mock_data_access)
+        self.scanner.service_config = mock_service_config
+
+        generate_res_for_0()
+        all_lifecycle_info = self.scanner._retrieve()
+        all_violations = self.scanner._find_violations(all_lifecycle_info)
+
+        res_map = {}
+        for i in _fake_bucket_list_0:
+            res_map[i.id] = i
+
+        expected_violations = set([
+            get_expect_violation_item(res_map, 'bkt1',
+                                      'exact retention 365', 0),
+            get_expect_violation_item(res_map, 'bkt2',
+                                      'exact retention 365', 0),
+            get_expect_violation_item(res_map, 'bkt3',
+                                      'exact retention 365', 0),
+            get_expect_violation_item(res_map, 'bkt4',
+                                      'exact retention 365', 0),
+            get_expect_violation_item(res_map, 'bkt5',
+                                      'exact retention 365', 0),
+            get_expect_violation_item(res_map, 'bkt6',
+                                      'exact retention 365', 0),
+            get_expect_violation_item(res_map, 'bkt7',
+                                      'exact retention 365', 0),
+        ])
+
+        self.assertEqual(expected_violations, set(all_violations))
+
+    @unittest.skipIf(do_not_test_old_cases, 'debug new test cases')
+    def test_retention_retrieve_22(self):
+        """test_retention_retrieve_22"""
+
+        rules_local_path = get_datafile_path(
+            __file__,
+            'bucket_retention_test_rules_22.yaml')
+        self.scanner = retention_scanner.RetentionScanner(
+            {}, {}, mock.MagicMock(), '', '', rules_local_path)
+
+        mock_data_access = mock.MagicMock()
+        mock_data_access.scanner_iter.side_effect = _mock_gcp_resource_iter_22
+        mock_service_config = mock.MagicMock()
+        mock_service_config.model_manager = mock.MagicMock()
+        mock_service_config.model_manager.get.return_value = (
+            mock.MagicMock(), mock_data_access)
+        self.scanner.service_config = mock_service_config
+
+        generate_res_for_22()
+        all_lifecycle_info = self.scanner._retrieve()
+        all_violations = self.scanner._find_violations(all_lifecycle_info)
+
+        res_map = {}
+        for i in _fake_bucket_list_22:
+            res_map[i.id] = i
+
+        expected_violations = set([
```

Comment:
Break this up into multiple tests so that each test only needs a small number of test resources, and produces a small number (or no) violations.